lsp-mode.el

```diff
@@ -5646,29 +5646,32 @@ perform the request synchronously."
            (lsp-request "workspace/symbol" `(:query ,pattern))))
 
 (defun lsp--get-symbol-to-rename ()
-  "Get symbol to rename and placeholder at point."
-  (if (let ((rename-provider (or (lsp--capability :renameProvider)
-                                 (-some-> (lsp--registered-capability "textDocument/rename")
-                                   (lsp--registered-capability-options)))))
-        (lsp:rename-options-prepare-provider? rename-provider))
-      (-when-let (response (lsp-request "textDocument/prepareRename"
-                                        (lsp--text-document-position-params)))
+  "Get symbol to rename and placeholder at point.
+Returns a cons (SYMBOL-TO-RENAME . PLACEHOLDER). Returns nil if
+renaming is supported but cannot be done at point."
+  (unless (lsp-feature? "textDocument/rename")
+    (error "The connected server(s) doesn't support renaming"))
+  (if (lsp-feature? "textDocument/prepareRename")
+      (when-let ((response
+                  (lsp-request "textDocument/prepareRename"
+                               (lsp--text-document-position-params))))
         (-let* (((start . end) (lsp--range-to-region
                                 (if (lsp-range? response)
                                     response
                                   (lsp:prepare-rename-result-range response))))
-                (symbol (buffer-substring-no-properties start end))
+                (symbol (buffer-substring start end))
                 (placeholder (lsp:prepare-rename-result-placeholder response)))
           (cons symbol (or placeholder symbol))))
-    (let ((symbol (thing-at-point 'symbol t)))
-      (cons symbol symbol))))
+    (when-let ((sym (thing-at-point 'symbol)))
+      (cons sym sym))))
 
 (defun lsp-rename (newname)
```

Comment:
Why don't we just check the capability in this command (public) instead of changing internal one?
I don't think guarding every internal one is desirable/good from perf POV. Just guard once in public one should be enough, shouldn't it?

Reply:
If we guard only in `lsp-rename`, the user would be prompted to enter a new name even though renaming is unsupported.

Reply:
Also, it's fast enough:
```emacs-lisp
(with-current-buffer "main.c"
  (benchmark-run 1000 (ignore (lsp-feature? "textDocument/rename"))))
```
yields 0.037 seconds.

Calling it only twice gives me 2ms timings, which is less than a 120FPS monitor's refresh latency.

Reply:
> If we guard only in `lsp-rename`, the user would be prompted to enter a new name even though renaming is unsupported.

see https://github.com/emacs-lsp/lsp-mode/pull/2317#issuecomment-724147973
https://github.com/yyoncho/lsp-mode/commit/443f801a4117361d2247c408b02dcc0d8c95a441 .

Reply:
Is it better to move the check outside of `lsp--get-symbol-to-rename` though?

Reply:
It's actually called only once though, in `lsp--get-symbol-to-rename`. We needn't add checks in every function, because the server will respond with a proper "not implemented" error if it isn't supported. Here, we cannot rely on that though, since we mustn't prompt the user for a name if the rename wouldn't work anyway.

Reply:
I think you don't get my point. Let say if the `lsp--get-symbol-to-rename` is gone/refactored, which can happen since it's internal function, now we will need to find a new place to put the capacity checking.
On other hand, `lsp-rename` is public function, and is unlikely to change without a breaking change, put the capacity guard there is the best, since user has to use that function to actually go through rename.
Is there any specific concern why the capacity guard has to be put in `lsp--get-symbol-to-rename`?

Reply:
We could add an additional `lsp-feature?` check to `lsp-rename`. However, the check in `lsp--get-symbol-to-rename` must stay: the interactive block is executed before the body, meaning the user would get prompted for the symbol to rename even though rename might not be possible at all or at point.

Reply:
You can put the check in `(interactive)` form too. Like what @yyoncho suggested.
If you hate that, you can even make `(interactive '(nil))` and check for `newName` outside of `interactive` block and call to `lsp--get-symbol-to-rename`.